File: ios/MitiMaiti/ViewModels/AuthViewModel.swift
import SwiftUI
import Combine

@MainActor
class AuthViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var hasCompletedOnboarding = false
    @Published var phone = ""
    /// Dial code from the country picker. The backend requires E.164
    /// (+<country><number>) — sending the bare national number 400s every
    /// phone OTP request.
    @Published var countryCode = "+91"
    @Published var email = ""

    /// Full E.164 number composed from the picker's dial code and the typed
    /// number. A typed leading + is trusted as a complete number.
    var e164Phone: String {
        let trimmed = phone.trimmingCharacters(in: .whitespaces)
        if trimmed.hasPrefix("+") { return "+" + trimmed.filter { $0.isNumber } }
        var digits = trimmed.filter { $0.isNumber }
        while digits.hasPrefix("0") { digits.removeFirst() }
        return countryCode + digits
    }
    @Published var otpCode = ""
    @Published var isLoading = false
    @Published var error: String?
    @Published var otpSent = false
    @Published var resendCooldown = 0
    @Published var resendCount = 0

    // Account linking (post-signup "secure your account" step + Settings)
    @Published var showLinkStep = false
    /// true → the link step asks for a mobile number (email/Google/Apple-first
    /// signups); false → it asks for an email (phone-first signups).
    @Published var linkStepSecuresPhone = false
    @Published var linkInProgress = false
    /// nil = idle, "success" = linked, otherwise a user-facing error message.
    @Published var linkResult: String? = nil
    @Published var linkEmailOtpSent = false
    @Published var pendingLinkEmail = ""

    private let api = APIService.shared
    private var timer: Timer?

    func sendOTP() {
        guard phone.count >= 10 else {
            error = "Please enter a valid phone number"
            return
        }
        isLoading = true
        error = nil

        Task {
            do {
                let success = try await api.sendOTP(phone: e164Phone)
                isLoading = false
                if success {
                    otpSent = true
                    resendCount += 1
                    startResendTimer()
                }
            } catch {
                isLoading = false
                self.error = error.localizedDescription
            }
        }
    }

    func verifyOTP() {
        guard otpCode.count == 6 else {
            error = "Please enter the 6-digit code"
            return
        }
        isLoading = true
        error = nil

        Task {
            do {
                let result = try await api.verifyOTP(phone: e164Phone, code: otpCode)
                isLoading = false
                hasCompletedOnboarding = !result.needsOnboarding
                // a phone-first signup already has its verified
                // phone — straight to onboarding, no link step. Email/Google
                // can be added later in Settings.
                showLinkStep = false
                isAuthenticated = true
                SocketChat.shared.connect(token: result.accessToken)
            } catch {
                isLoading = false
                self.error = error.localizedDescription
            }
        }
    }

    func sendEmailOTP() {
        guard isValidEmail(email) else {
            error = "Please enter a valid email address"
            return
        }
        isLoading = true
        error = nil

        Task {
            do {
                let success = try await api.sendEmailOTP(email: email)
                isLoading = false
                if success {
                    otpSent = true
                    resendCount += 1
                    startResendTimer()
                }
            } catch {
                isLoading = false
                self.error = "Failed to send code. Please try again."
            }
        }
    }

    func verifyEmailOTP() {
        guard otpCode.count == 6 else {
            error = "Please enter the 6-digit code"
            return
        }
        isLoading = true
        error = nil

        Task {
            do {
                let result = try await api.verifyEmailOTP(email: email, code: otpCode)
                isLoading = false
                hasCompletedOnboarding = !result.needsOnboarding
                // New email-first users get the "secure your mobile number" step.
                showLinkStep = result.needsOnboarding
                linkStepSecuresPhone = true
                isAuthenticated = true
                SocketChat.shared.connect(token: result.accessToken)
            } catch {
                isLoading = false
                self.error = "Invalid code. Please try again."
            }
        }
    }

    func clearLinkResult() { linkResult = nil }
    func finishLinkStep() { showLinkStep = false; linkResult = nil }

    func linkEmailAddress(_ rawEmail: String) {
        let email = rawEmail.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard email.contains("@"), email.contains(".") else {
            linkResult = "Please enter a valid email address"; return
        }
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                try await api.linkEmail(email)
                linkInProgress = false
                linkResult = "success"
            } catch {
                linkInProgress = false
                linkResult = "Couldn't link that email — it may already be in use."
            }
        }
    }

    /// Step 1: send an OTP to the email the user wants to add.
    func linkEmailStart(_ rawEmail: String) {
        let email = rawEmail.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard email.contains("@"), email.contains(".") else {
            linkResult = "Please enter a valid email address"; return
        }
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                try await api.linkEmailStart(email)
                pendingLinkEmail = email
                linkEmailOtpSent = true
                linkInProgress = false
            } catch {
                linkInProgress = false
                linkResult = "Couldn't send the code. Please try again."
            }
        }
    }

    /// Step 2: verify the emailed code → attaches the email (or auto-merges).
    func linkEmailVerify(_ code: String) {
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                let merged = try await api.linkEmailVerify(pendingLinkEmail, code: code)
                linkInProgress = false
                linkResult = merged ? "merged" : "success"
            } catch {
                linkInProgress = false
                linkResult = "That code is invalid or expired."
            }
        }
    }

    func resetLinkEmailOtp() { linkEmailOtpSent = false; pendingLinkEmail = "" }

    func linkGoogleAccount() {
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                let idToken = try await GoogleSignInService.signIn()
                // Prefill the onboarding name from the ID token when we don't
                // have one yet (phone-OTP signup linking Google before
                // onboarding). Never overwrites an existing name.
                if UserProfileStore.shared.firstName.isEmpty,
                   let name = Self.nameFromIdToken(idToken), !name.isEmpty {
                    UserProfileStore.shared.firstName = name
                }
                // Google OAuth is trusted directly — no email OTP.
                let merged = try await api.linkGoogle(idToken: idToken)
                linkInProgress = false
                linkResult = merged ? "merged" : "success"
            } catch GoogleSignInError.canceled {
                linkInProgress = false
            } catch {
                linkInProgress = false
                linkResult = "Couldn't link Google. Please try again."
            }
        }
    }

    // MARK: - Phone linking (email-first accounts securing a mobile number)

    @Published var linkPhoneOtpSent = false
    @Published var pendingLinkPhone = ""

    /// Step 1: send an SMS OTP to the phone the user wants to add.
    func linkPhoneStart(_ rawPhone: String) {
        let phone = rawPhone.filter { !$0.isWhitespace }
        guard phone.range(of: #"^\+[1-9]\d{6,14}$"#, options: .regularExpression) != nil else {
            linkResult = "Enter the number with country code, e.g. +971501234567"; return
        }
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                try await api.linkPhoneStart(phone)
                pendingLinkPhone = phone
                linkPhoneOtpSent = true
                linkInProgress = false
            } catch {
                linkInProgress = false
                linkResult = "Couldn't send the code. Please try again."
            }
        }
    }

    /// Step 2: verify the SMS code → attaches the phone (or auto-merges).
    func linkPhoneVerify(_ code: String) {
        linkInProgress = true
        linkResult = nil
        Task {
            do {
                let merged = try await api.linkPhoneVerify(pendingLinkPhone, code: code)
                linkInProgress = false
                linkResult = merged ? "merged" : "success"
            } catch {
                linkInProgress = false
                linkResult = "That code is invalid or expired."
            }
        }
    }

    func resetLinkPhoneOtp() { linkPhoneOtpSent = false; pendingLinkPhone = "" }

    func signInWithGoogle(idToken: String) {
        isLoading = true
        error = nil

        // Extract the user's name from the Google ID token *before* we hit the
        // backend, so onboarding prefill works even if the backend response
        // doesn't carry it back. The ID token is a JWT and its payload
        // contains the `name` claim (full name) plus `given_name`.
        if let name = Self.nameFromIdToken(idToken), !name.isEmpty {
            UserProfileStore.shared.firstName = name
        }

        Task {
            do {
                let result = try await api.verifyGoogleIdToken(idToken)
                isLoading = false
                if let firstName = result.firstName, !firstName.isEmpty {
                    UserProfileStore.shared.firstName = firstName
                }
                hasCompletedOnboarding = !result.needsOnboarding
                // New Google-first users get the "secure your mobile number" step.
                showLinkStep = result.needsOnboarding
                linkStepSecuresPhone = true
                isAuthenticated = true
                SocketChat.shared.connect(token: result.accessToken)
            } catch {
                isLoading = false
                self.error = "Google sign-in failed: \(error.localizedDescription)"
            }
        }
    }

    /// Pull the local-part of the `email` claim out of an Apple ID token JWT
    /// and capitalize it as a best-effort name. Apple doesn't expose a name
    /// claim in the token; this is a fallback when credential.fullName is nil
    /// (common after the first sign-in). For `karan.narwani@example.com` this
    /// returns "Karan Narwani"; for `karannarwani01@hotmail.com" it returns
    /// "Karannarwani01" — not great, but better than empty.
    private static func nameGuessFromAppleIdToken(_ idToken: String) -> String? {
        let parts = idToken.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
        let pad = (4 - payload.count % 4) % 4
        payload += String(repeating: "=", count: pad)
        payload = payload.replacingOccurrences(of: "-", with: "+")
                          .replacingOccurrences(of: "_", with: "/")
        guard let data = Data(base64Encoded: payload),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let email = json["email"] as? String,
              let localPart = email.split(separator: "@").first else {
            return nil
        }
        // Strip trailing digits — `karannarwani01` → `karannarwani` — so the
        // prefill reads as a name, not a username.
        let cleaned = String(localPart).trimmingCharacters(in: .decimalDigits)
        guard !cleaned.isEmpty else { return nil }
        return cleaned
            .split(whereSeparator: { ".+_-".contains($0) })
            .map { $0.prefix(1).uppercased() + $0.dropFirst() }
            .joined(separator: " ")
    }

    /// Decode the `name` (or `given_name`) claim from a Google ID token JWT
    /// without verifying the signature — we only trust this for UI prefill,
    /// not for auth.
    private static func nameFromIdToken(_ idToken: String) -> String? {
        let parts = idToken.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
        // JWT uses base64url; pad to a multiple of 4 for Data(base64Encoded:).
        let pad = (4 - payload.count % 4) % 4
        payload += String(repeating: "=", count: pad)
        payload = payload.replacingOccurrences(of: "-", with: "+")
                          .replacingOccurrences(of: "_", with: "/")
        guard let data = Data(base64Encoded: payload),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        if let name = json["name"] as? String, !name.isEmpty { return name }
        if let given = json["given_name"] as? String, !given.isEmpty { return given }
        return nil
    }

    func setGoogleSignInError(_ message: String) {
        error = message
    }

    func signInWithApple(
        idToken: String,
        nonce: String?,
        givenName: String?,
        familyName: String?
    ) {
        isLoading = true
        error = nil

        // Apple only sends fullName on the very first sign-in for a given
        // Apple ID and even then is inconsistent — on subsequent sign-ins
        // both fields come back nil. Use fullName when we have it; otherwise
        // fall back to the email local-part from the ID token JWT so the
        // onboarding "What's your full name?" field still has *something* to
        // prefill instead of staying blank.
        let appleName = [givenName, familyName].compactMap { $0 }.joined(separator: " ")
        if !appleName.isEmpty {
            UserProfileStore.shared.firstName = appleName
        } else if let fallback = Self.nameGuessFromAppleIdToken(idToken), !fallback.isEmpty {
            UserProfileStore.shared.firstName = fallback
        }

        Task {
            do {
                let result = try await api.verifyAppleIdToken(
                    idToken,
                    nonce: nonce,
                    givenName: givenName,
                    familyName: familyName
                )
                isLoading = false
                if let firstName = result.firstName, !firstName.isEmpty {
                    UserProfileStore.shared.firstName = firstName
                }
                hasCompletedOnboarding = !result.needsOnboarding
                // New Apple-first users get the "secure your mobile number" step.
                showLinkStep = result.needsOnboarding
                linkStepSecuresPhone = true
                isAuthenticated = true
                SocketChat.shared.connect(token: result.accessToken)
            } catch {
                isLoading = false
                self.error = "Apple sign-in failed. Please try again."
            }
        }
    }

    func setAppleSignInError(_ message: String) {
        error = message
    }

    func resetOtpState() {
        otpSent = false
        otpCode = ""
        error = nil
    }

    func completeOnboarding() {
        hasCompletedOnboarding = true
    }

    /// Attempt to resume a previously saved session on app launch. If valid
    /// tokens are found and accepted by the backend, the user skips the
    /// Welcome screen; onboarding is shown only if the server says it's still
    /// needed. No stored/valid session leaves the user on Welcome.
    func restoreSession() async {
        guard let needsOnboarding = await api.restoreSession() else { return }
        hasCompletedOnboarding = !needsOnboarding
        isAuthenticated = true
        if let token = await api.currentAccessToken() {
            SocketChat.shared.connect(token: token)
        }
    }

    func logout() {
        isAuthenticated = false
        hasCompletedOnboarding = false
        phone = ""
        email = ""
        otpCode = ""
        otpSent = false
        SocketChat.shared.disconnect()
        Task {
            await api.clearTokens()
        }
    }

    private func isValidEmail(_ s: String) -> Bool {
        let pattern = #"^[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$"#
        return s.range(of: pattern, options: .regularExpression) != nil
    }

    private func startResendTimer() {
        resendCooldown = 30
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] t in
            Task { @MainActor in
                guard let self else { t.invalidate(); return }
                if self.resendCooldown > 0 {
                    self.resendCooldown -= 1
                } else {
                    t.invalidate()
                }
            }
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/AuthViewModel.kt
package com.mitimaiti.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.User
import com.mitimaiti.app.services.APIError
import com.mitimaiti.app.services.APIService
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel : ViewModel() {
    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()
    private val _hasCompletedOnboarding = MutableStateFlow(false)
    val hasCompletedOnboarding: StateFlow<Boolean> = _hasCompletedOnboarding.asStateFlow()
    private val _phone = MutableStateFlow("")
    val phone: StateFlow<String> = _phone.asStateFlow()
    private val _email = MutableStateFlow("")
    val email: StateFlow<String> = _email.asStateFlow()
    private val _otpCode = MutableStateFlow("")
    val otpCode: StateFlow<String> = _otpCode.asStateFlow()
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    private val _otpSent = MutableStateFlow(false)
    val otpSent: StateFlow<Boolean> = _otpSent.asStateFlow()
    private val _resendCooldown = MutableStateFlow(0)
    val resendCooldown: StateFlow<Int> = _resendCooldown.asStateFlow()
    private val _resendCount = MutableStateFlow(0)
    val resendCount: StateFlow<Int> = _resendCount.asStateFlow()
    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    // Account linking (post-OTP "add a backup sign-in" screen + Settings)
    private val _linkInProgress = MutableStateFlow(false)
    val linkInProgress: StateFlow<Boolean> = _linkInProgress.asStateFlow()
    /** null = idle, "success" = linked, otherwise a user-facing error message. */
    private val _linkResult = MutableStateFlow<String?>(null)
    val linkResult: StateFlow<String?> = _linkResult.asStateFlow()
    /** true once the email OTP has been sent → show the code step. */
    private val _linkEmailOtpSent = MutableStateFlow(false)
    val linkEmailOtpSent: StateFlow<Boolean> = _linkEmailOtpSent.asStateFlow()
    private val _pendingLinkEmail = MutableStateFlow("")
    val pendingLinkEmail: StateFlow<String> = _pendingLinkEmail.asStateFlow()

    fun updatePhone(value: String) { _phone.value = value }
    fun updateEmail(value: String) { _email.value = value.trim() }
    fun updateOtpCode(value: String) { _otpCode.value = value.take(6) }
    fun clearError() { _error.value = null }
    fun resetOtpState() { _otpSent.value = false; _otpCode.value = ""; _error.value = null }
    fun clearLinkResult() { _linkResult.value = null }

    fun linkEmail(rawEmail: String) {
        val email = rawEmail.trim()
        if (!isValidEmail(email)) { _linkResult.value = "Please enter a valid email address"; return }
        viewModelScope.launch {
            _linkInProgress.value = true
            APIService.linkEmail(email)
                .onSuccess { _linkResult.value = "success" }
                .onFailure {
                    _linkResult.value = when (it) {
                        is APIError.LinkConflict -> "That email is already linked to another account"
                        else -> "Couldn't link that email. Please try again."
                    }
                }
            _linkInProgress.value = false
        }
    }

    /** Step 1: send an OTP to the email the user wants to add. */
    fun linkEmailStart(rawEmail: String) {
        val email = rawEmail.trim()
        if (!isValidEmail(email)) { _linkResult.value = "Please enter a valid email address"; return }
        viewModelScope.launch {
            _linkInProgress.value = true
            APIService.linkEmailStart(email)
                .onSuccess { _pendingLinkEmail.value = email; _linkEmailOtpSent.value = true }
                .onFailure { _linkResult.value = "Couldn't send the code. Please try again." }
            _linkInProgress.value = false
        }
    }

    /** Step 2: verify the emailed code → attaches the email (or auto-merges). */
    fun linkEmailVerify(code: String) {
        viewModelScope.launch {
            _linkInProgress.value = true
            APIService.linkEmailVerify(_pendingLinkEmail.value, code)
                .onSuccess { merged -> _linkResult.value = if (merged) "merged" else "success" }
                .onFailure {
                    _linkResult.value = when (it) {
                        is APIError.InvalidOtp -> "That code is invalid or expired"
                        is APIError.LinkConflict -> "That email belongs to another active account"
                        else -> "Couldn't verify the code. Please try again."
                    }
                }
            _linkInProgress.value = false
        }
    }

    fun resetLinkEmailOtp() { _linkEmailOtpSent.value = false; _pendingLinkEmail.value = "" }

    fun linkGoogle(idToken: String) {
        // Prefill the onboarding name from the Google ID token when we don't
        // have one yet (phone-OTP signup linking Google before onboarding).
        // Never overwrites a name the user already has.
        if (com.mitimaiti.app.services.UserPrefs.firstName.value.isBlank()) {
            nameFromIdToken(idToken)?.takeIf { it.isNotBlank() }?.let {
                com.mitimaiti.app.services.UserPrefs.setFirstName(it)
            }
        }
        viewModelScope.launch {
            _linkInProgress.value = true
            // Google OAuth is trusted directly — no email OTP.
            APIService.linkGoogle(idToken)
                .onSuccess { merged -> _linkResult.value = if (merged) "merged" else "success" }
                .onFailure {
                    _linkResult.value = when (it) {
                        is APIError.LinkConflict -> "That Google account is already linked elsewhere"
                        else -> "Couldn't link Google. Please try again."
                    }
                }
            _linkInProgress.value = false
        }
    }

    /** true once the phone OTP has been sent → show the code step. */
    private val _linkPhoneOtpSent = MutableStateFlow(false)
    val linkPhoneOtpSent: StateFlow<Boolean> = _linkPhoneOtpSent.asStateFlow()
    private val _pendingLinkPhone = MutableStateFlow("")
    val pendingLinkPhone: StateFlow<String> = _pendingLinkPhone.asStateFlow()

    /** Step 1: send an SMS OTP to the phone the user wants to add. */
    fun linkPhoneStart(rawPhone: String) {
        val phone = rawPhone.filter { !it.isWhitespace() }
        if (!phone.matches(Regex("^\\+[1-9]\\d{6,14}$"))) {
            _linkResult.value = "Enter the number with country code, e.g. +971501234567"
            return
        }
        viewModelScope.launch {
            _linkInProgress.value = true
            APIService.linkPhoneStart(phone)
                .onSuccess { _pendingLinkPhone.value = phone; _linkPhoneOtpSent.value = true }
                .onFailure { _linkResult.value = "Couldn't send the code. Please try again." }
            _linkInProgress.value = false
        }
    }

    /** Step 2: verify the SMS code → attaches the phone (or auto-merges). */
    fun linkPhoneVerify(code: String) {
        viewModelScope.launch {
            _linkInProgress.value = true
            APIService.linkPhoneVerify(_pendingLinkPhone.value, code)
                .onSuccess { merged -> _linkResult.value = if (merged) "merged" else "success" }
                .onFailure {
                    _linkResult.value = when (it) {
                        is APIError.InvalidOtp -> "That code is invalid or expired"
                        is APIError.LinkConflict -> "That number belongs to another active account"
                        else -> "Couldn't verify the code. Please try again."
                    }
                }
            _linkInProgress.value = false
        }
    }

    fun resetLinkPhoneOtp() { _linkPhoneOtpSent.value = false; _pendingLinkPhone.value = "" }

    fun sendOTP() {
        if (_phone.value.length < 10) { _error.value = "Please enter a valid phone number"; return }
        viewModelScope.launch {
            _isLoading.value = true; _error.value = null
            APIService.sendOTP(_phone.value).onSuccess { _otpSent.value = true; _resendCount.value++; startResendTimer() }.onFailure {
                _error.value = (it as? APIError.MessageRejected)?.reason ?: "Failed to send OTP. Please try again."
            }
            _isLoading.value = false
        }
    }

    fun verifyOTP() {
        if (_otpCode.value.length != 6) { _error.value = "Please enter a 6-digit code"; return }
        viewModelScope.launch {
            _isLoading.value = true; _error.value = null
            APIService.verifyOTP(_phone.value, _otpCode.value).onSuccess { (user, isNewUser) ->
                _currentUser.value = user
                _isAuthenticated.value = true
                // Route on the server's authoritative needs_onboarding flag
                // (mirrors iOS). isNewUser guards the case where the flag is
                // absent for a brand-new account.
                _hasCompletedOnboarding.value = !isNewUser && !user.needsOnboarding
            }.onFailure { _error.value = "Invalid OTP. Please try again." }
            _isLoading.value = false
        }
    }

    fun sendEmailOTP() {
        if (!isValidEmail(_email.value)) { _error.value = "Please enter a valid email address"; return }
        viewModelScope.launch {
            _isLoading.value = true; _error.value = null
            APIService.sendEmailOTP(_email.value).onSuccess { _otpSent.value = true; _resendCount.value++; startResendTimer() }
                .onFailure { _error.value = "Failed to send code. Please try again." }
            _isLoading.value = false
        }
    }

    fun verifyEmailOTP() {
        if (_otpCode.value.length != 6) { _error.value = "Please enter a 6-digit code"; return }
        viewModelScope.launch {
            _isLoading.value = true; _error.value = null
            APIService.verifyEmailOTP(_email.value, _otpCode.value).onSuccess { (user, isNewUser) ->
                _currentUser.value = user
                _isAuthenticated.value = true
                // Route on the server's authoritative needs_onboarding flag
                // (mirrors iOS). isNewUser guards the case where the flag is
                // absent for a brand-new account.
                _hasCompletedOnboarding.value = !isNewUser && !user.needsOnboarding
            }.onFailure { _error.value = "Invalid code. Please try again." }
            _isLoading.value = false
        }
    }

    fun signInWithGoogle(idToken: String) {
        // Prefill the onboarding name field locally from the ID token JWT
        // before we even talk to the backend — works even when the response
        // doesn't echo a name.
        nameFromIdToken(idToken)?.takeIf { it.isNotBlank() }?.let {
            com.mitimaiti.app.services.UserPrefs.setFirstName(it)
        }
        viewModelScope.launch {
            _isLoading.value = true; _error.value = null
            APIService.verifyGoogleIdToken(idToken).onSuccess { (user, isNewUser) ->
                _currentUser.value = user
                _isAuthenticated.value = true
                // Route on the server's authoritative needs_onboarding flag
                // (mirrors iOS). isNewUser guards the case where the flag is
                // absent for a brand-new account.
                _hasCompletedOnboarding.value = !isNewUser && !user.needsOnboarding
            }.onFailure { _error.value = "Google sign-in failed. Please try again." }
            _isLoading.value = false
        }
    }

    /**
     * Decode the `name` (or `given_name`) claim out of a Google ID token JWT.
     * Signature is not verified — only used for UI prefill, never for auth.
     */
    private fun nameFromIdToken(idToken: String): String? {
        return try {
            val parts = idToken.split('.')
            if (parts.size < 2) return null
            val payload = android.util.Base64.decode(
                parts[1],
                android.util.Base64.URL_SAFE or android.util.Base64.NO_WRAP or android.util.Base64.NO_PADDING
            )
            val json = org.json.JSONObject(String(payload, Charsets.UTF_8))
            (json.optString("name").takeIf { it.isNotBlank() })
                ?: json.optString("given_name").takeIf { it.isNotBlank() }
        } catch (e: Exception) { null }
    }

    fun setGoogleSignInError(message: String) {
        _error.value = message
    }

    private fun isValidEmail(s: String): Boolean =
        s.isNotBlank() && android.util.Patterns.EMAIL_ADDRESS.matcher(s).matches()

    fun completeOnboarding() { _hasCompletedOnboarding.value = true }
    fun logout() { _isAuthenticated.value = false; _hasCompletedOnboarding.value = false; _phone.value = ""; _email.value = ""; _otpCode.value = ""; _otpSent.value = false; _currentUser.value = null; APIService.clearTokens() }

    /**
     * Hydrate auth state from a session that was already in token storage when
     * the app launched, so the user lands directly on Main / Onboarding instead
     * of bouncing through Welcome.
     */
    fun bootstrapAuthenticated(user: User, hasOnboarded: Boolean) {
        _currentUser.value = user
        _isAuthenticated.value = true
        _hasCompletedOnboarding.value = hasOnboarded
    }

    private fun startResendTimer() {
        viewModelScope.launch { _resendCooldown.value = 30; while (_resendCooldown.value > 0) { delay(1000); _resendCooldown.value-- } }
    }
}
